### PR TITLE
Modify speaker input

### PR DESCRIPTION
**Pull request by nithinraok**
This PR brings a major revamp to speaker diarization inference, with following are notable changes:
* Input to offline_diarization.py and offline_diarization_with_asr.py expects manifest file ( Now no longer need to create individual files for audio paths, rttms, oracle speakers and uem filepaths)
* With oracle_vad option, CLusteringDiarizer automatically computes oracle_manifest without being fed as the separate manifest file
* Updated yaml file and is now more clear with many options for VAD, Speaker Embeddings and Clustering 
* Separated clustering and scoring modules
* Improved pyannote usage with uris passed as uniq_key values
* Updated Speaker_Diairization_Inference.ipynb and ASR_with_Speaker_Diarization.ipynb 
* Removed duplicate functions from diarization utils and avoided recalculation of diarization score 
* Added Jenkins Test for diarization with asr 
* Updated README and DocStrings

### CODE REVIEW

**Review by tango4j — DISMISSED** [review]

File: nemo/collections/asr/parts/utils/diarization_utils.py [hunk 1]
```diff
@@ -604,80 +560,31 @@ def get_word_ts_from_spaces(self, char_ts, _spaces_in_sec, end_stamp):
...
-        if oracle_num_speakers != None:
-            if oracle_num_speakers.isnumeric():
-                oracle_num_speakers = int(oracle_num_speakers)
-            elif oracle_num_speakers in NONE_LIST:
-                oracle_num_speakers = None
+        if diar_model_config.diarizer.asr.parameters.asr_based_vad:
+            self.save_VAD_labels_list(words_and_timestamps, self.audio_file_list)
+            oracle_manifest = os.path.join(self.root_path, 'asr_vad_manifest.json')
+            oracle_manifest = write_rttm2manifest(self.VAD_RTTM_MAP, oracle_manifest)
+            diar_model_config.diarizer.vad.model_path = None
```
[the remark this answers is not in the document]
Comment (nithinraok) [reply nemo/collections/asr/parts/utils/diarization_utils.py:576]: I see, in tutorial same config was used to perform both vad and oracle vad computation, hence the issue. Thanks for updating this.

**Review by tango4j — DISMISSED** [review]

File: nemo/collections/asr/parts/utils/diarization_utils.py [hunk 2]
```diff
@@ -686,104 +592,68 @@ def get_frame_level_VAD(self, oracle_model, audio_file_list):
...
                 List contains file paths for audio files.
         """
-        for k, audio_file_path in enumerate(audio_file_list):
-            uniq_id = get_uniq_id_from_audio_path(audio_file_path)
-            vad_pred_diar = oracle_model.vad_pred_dir
-            frame_vad = os.path.join(vad_pred_diar, uniq_id + '.median')
-            frame_vad_list = read_file_paths(frame_vad)
-            frame_vad_float_list = [float(x) for x in frame_vad_list]
+        for uniq_id in self.AUDIO_RTTM_MAP:
+            frame_vad = os.path.join(vad_processing_dir, uniq_id + '.median')
```
Comment (fayejf) [inline nemo/collections/asr/parts/utils/diarization_utils.py:596]: .median 
Will need to update the median here later. the ext could be either median or mean or frame as the user chose.

File: nemo/collections/asr/parts/utils/diarization_utils.py [hunk 3]
```diff
@@ -686,104 +592,68 @@ def get_frame_level_VAD(self, oracle_model, audio_file_list):
...
                 "DER": DER,
                 "CER": CER,
                 "FA": FA,
                 "MISS": MISS,
                 "n_spk": est_n_spk,
-                "mapping": mapping[0],
+                "mapping": mapping_dict[key],
                 "spk_counting": (est_n_spk == ref_n_spk),
             }
+            logging.info("score for key {}: {}".format(key, DER_result_dict[key]))
```
Comment (fayejf) [inline nemo/collections/asr/parts/utils/diarization_utils.py:639]: key? Maybe _sample_ or _session_?

**Review by fayejf — DISMISSED** [review]
This is awesome! Thank you!

File: examples/speaker_tasks/diarization/README.md [hunk 4]
```diff
@@ -30,55 +30,38 @@ Diarization Error Rate (DER) table of `ecapa_tdnn.nemo` model on well known eval
...
 ```
-&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; `diarizer.oracle_num_speakers='number_of_speakers.list'` 
+Mandatory fields are `audio_filepath`, `offset`, `duration`, `label:"infer"` and `text: <ground truth or "-" >`  , and the rest are optional keys which can be passed based on the type of evaluation 
 
-&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; If all sessions have the same number of speakers, you can specify the option as: 
-
-&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; `diarizer.oracle_num_speakers=5`
+Some of important options in config file: 
 
 - **`diarizer.speaker_embeddings.model_path`: speaker embedding model name**
```
[the remark this answers is not in the document]
Comment (nithinraok) [reply examples/speaker_tasks/diarization/README.md:67]: Looks okay to me... https://catalog.ngc.nvidia.com/models?filters=&orderBy=scoreDESC&pageNumber=0&query=SpeakerNet&quickFilter=

File: examples/speaker_tasks/diarization/README.md [hunk 5]
```diff
@@ -30,55 +30,38 @@ Diarization Error Rate (DER) table of `ecapa_tdnn.nemo` model on well known eval
```
[the remark this answers is not in the document]
Comment (nithinraok) [reply examples/speaker_tasks/diarization/README.md:30]: TitaNet model isn't published yet, Will update this and also at many other places once its published

File: examples/speaker_tasks/diarization/conf/offline_diarization_with_asr.yaml [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+      oracle_num_speakers: False # if True, uses num of speakers value provided in manifest file
+      max_num_speakers: 20 # max number of speakers for each recording. If oracle num speakers is passed this value is ignored.
+      enhanced_count_thres: 80
+      max_rp_threshold: 0.25
+      sparse_search_volume: 30
+  
+  asr:
+    model_path: ??? # Provie NGC cloud ASR model name
+    parameters:
+      asr_based_vad: False # if true, speech segmentation for diarization is based on word-timestamps from ASR inference.
```
[the remark this answers is not in the document]
Comment (nithinraok) [reply examples/speaker_tasks/diarization/conf/offline_diarization_with_asr.yaml:49]: So, idea is to separate asr from existing offline speaker diarization groups hence a separate section for asr and its related parameters.

File: examples/speaker_tasks/diarization/README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (fayejf) [reply examples/speaker_tasks/diarization/README.md:67]: Okay just thought we can have a better query/tag for searching the models...

File: examples/speaker_tasks/diarization/offline_diarization_with_asr.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+from omegaconf import OmegaConf
+
+from nemo.collections.asr.parts.utils.diarization_utils import ASR_DIAR_OFFLINE
+from nemo.collections.asr.parts.utils.speaker_utils import audio_rttm_map
+from nemo.core.config import hydra_runner
+from nemo.utils import logging
+
+
+"""
+This script demonstrates how to use run speaker diarization.
```
[the remark this answers is not in the document]
Comment (nithinraok) [reply examples/speaker_tasks/diarization/offline_diarization_with_asr.py:26]: Sorry Fei, could you mention what is typo here. I cant see it :P Oh, you mean grammar mistake. Will correct it

File: nemo/collections/asr/parts/utils/diarization_utils.py (hunk 3 above)
Comment (nithinraok) [reply nemo/collections/asr/parts/utils/diarization_utils.py:639]: Yeah session sounds good.

**Review by fayejf — APPROVED** [review]

**Review by fayejf — APPROVED** [review]

**Pull request merged**